org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend

```diff
@@ -320,13 +319,19 @@ import static org.eclipse.xtext.diagnostics.Severity.*
 		initialized.thenAccept([
 			val diagnostics = new PublishDiagnosticsParams => [
 				it.uri = toUriString(uri)
-				it.diagnostics = issues.filter[severity !== IGNORE].map[toDiagnostic].toList
+				if (issues.isEmpty) {
+					diagnostics = #[]
+				} else {
+					diagnostics = workspaceManager.doRead(uri) [document, resource|
+						issues.filter[severity !== IGNORE].map[toDiagnostic(document, it)].toList
```

Comment:
any idea how to do this better?

Reply:
Should we have correct methods on the issue itself, eg `getRegion():ITextRegionWithLineInformation` incl a default method that has at least the currently (broken) semantics as it was implemented here? Note that line number could either start with 0 or 1 - we may have to convert here accordingly.

Reply:
do we have some api  that helps to build that information in an efficient manner?

Reply:
=> we would have to change IssueLocation and LocationData too (we copy over the diags 3 times)

Reply:
> API

NodeModelUtils.getLineAndColumn(INode, int) should be very efficient.

> IssueLocation / IssueImpl

Yes, unfortunately :(

Reply:
This cleanup would make a good first issue I suppose. So I'd be ok with the solution that is implemented here. Should we create a follow up ticket?

Reply:
i will have a look what can be done. mayby adding an issueextension to issue instead of a default interface?

Reply:
If that simplifies things, I'd be perfectly fine with that, too.

Reply:
There are quite a few inconsistencies / mismatches between docs and impl. The refactoring was attempted in PR #1212 but postponed.